### PR TITLE
Build configuration migration .NET 4.7.1 & .NET Core 2.0

### PR DESCRIPTION
**Pull request by Latency**
Clean working directory to mirror Master layout.

Rewrote ResourceGenerator for async handling.
- Downloads files 50% faster with considerable file size reduction.
- Uses new Image format to replace Bitmap.  Compiliant with .NET Core.

Updated framework .NET 4.7,1 and libraries to .NET 4.7.1 & .NET Core 2.0.
Started work on HearthMirror for .NET Standard 2.0 + aforementioned.
Revamped Pre/Postevent.bat file to better automate & installation for resource generation.
Fixed various hotpoints and related NPEs throughout ALL assemblies.
Updated build configuration for VSTools 15 which eliminates problems for binding references.
Merged all other developer efforts during this past month as master.
Updated assembly manifest versioning info v1.5.11.   Compiler set to (Default) latest ... e.g. C#7.0.

Changed namespace locations for Hearthstone Deck Tracker xaml files and dependancies so they match directory hierarchy locations.  Important for ms-build configuration v15.

### CODE REVIEW

**Comment by azeier** [comment]
I'm not gonna look through a +4,018 −3,607 PR, sorry.

This looks like a bunch of unnecessary changes.

**Comment by Latency** [comment]
Major work here!  Took me about 3 weeks to refactor all this and make it more bulletproof.   Spent an enormous amount of time to get the build configuration right also, since the namespace for most of the xaml files was not matching the folder hierarchy.    I cherry picked all the stuff from the past month.
It is tested and working!    Which includes my new ResourceGenerator which is 2x more efficient.

**Comment by Latency** [comment]
I have no idea if I can even get the appveyor to validate it.  Its major work.   Can you look into it for me please?  I really don't want to have all this effort go to waste.

**Comment by azeier** [comment]
I don't think I'll have time to look through this (and the other 2? PRs) in the next two weeks. This should probably be multiple PRs with reasonable diffs. 

One obvious issue right now: no PR should include version increments.

**Comment by Latency** [comment]
Well I can revert the version increment and put it to the same patch rev.   Didn't think that would made sense since there was improvements made... but if you are the only one responsible for doing that, then I understand!

I am not sure how to go about this with the other things.

1.  Namespace refactoring.
2.  Hotpoint and NPE validation.
3.  ResourceGenerator upgrade.
4.  Build configuration upgrade  + library upgrade + Framework upgrade (multi-support) + signed asm.

   The refactoring for the namespaces was critical and extremely delegate. (1/4)
I was at this for days.    To start over with this is a challenge and very time consuming.  It was necessary to migrate to the new build configuration. (4/4)

The code analyzers for NPEs and rework was probably unnecessary, yet beneficial to the overall health of the project (2/4)

Everything is intertwined.   I could easily split the ResourceGenerator upgrade.  (3/4)

**Comment by beheh** [comment]
@Latency Huge parts of your commits seem to be just replacing the LF line endings with CRLF. Make sure your editors/git respect the line endings as set out in the .editorconfig (LF only). Otherwise the diffs become unreadable.

You can set up your git to use LF only by setting the [core.eol](https://git-scm.com/docs/git-config#git-config-coreeol) setting.

File: Hearthstone Deck Tracker/Controls/DeckPicker/DeckPickerItemLayouts/DeckPickerItemLayout1.xaml [hunk 1]
```diff
@@ -81,7 +81,7 @@
                 <DockPanel Margin="0,-2,0,0" Height="19">
                     <Viewbox DockPanel.Dock="Left" MinWidth="70" StretchDirection="DownOnly" Stretch="Fill"
                              HorizontalAlignment="Left">
-                        <TextBlock Text="{Binding StatsString, RelativeSource={RelativeSource AncestorType=deckPicker:DeckPickerItem}}" HorizontalAlignment="Left"
```
Comment (Zawodowiec1532) [inline Hearthstone Deck Tracker/Controls/DeckPicker/DeckPickerItemLayouts/DeckPickerItemLayout1.xaml]: :(

**Comment by Latency** [comment]
I have my line endings set to  check out as is.. commit as UNIX line endings.  Not sure what is going on there!

File: Hearthstone Deck Tracker/Controls/DeckPicker/DeckPickerItemLayouts/DeckPickerItemLayout1.xaml (hunk 1 above)
Comment (Latency) [reply Hearthstone Deck Tracker/Controls/DeckPicker/DeckPickerItemLayouts/DeckPickerItemLayout1.xaml]: Somebody recently had made these changes.. where they took out the RelativeSource=foo
I just merged it as is..  take it up with the other guy.  There were about 2-4 of these.

-  RelativeSource={RelativeSource AncestorType=deckPicker:DeckPickerItem}}"

Also there were a couple changes where  Margin="5" was removed on select .xaml files.
I only merged them!  Again.. not my doing!

Please advise if we need to revert back on those GUI changes?

 PS.

   I should have been up to date with merging everybody elses contributions.  I did not make changes to the GUIs.   In fact, other than the ResourceGenerator revamp and the refactoring of namespaces.. to match folder heirarchy..  everythign else should have been just optomizations with code analyzers and new format conventions as SAME!

The goal was to not change or break anything I didn't already have vetted.

**Comment by riQQ** [comment]
@beheh it's the other way round. This pr changes CR LF -> LF. There a quite some files that have completely wrong or mixed line endings.

File: Hearthstone Deck Tracker/Controls/DeckPicker/DeckPickerItemLayouts/DeckPickerItemLayout1.xaml (hunk 1 above)
Comment (Zawodowiec1532) [reply Hearthstone Deck Tracker/Controls/DeckPicker/DeckPickerItemLayouts/DeckPickerItemLayout1.xaml]: The RelativeSources were added, not deleted. Files which you are trying to push here are not updated. - In short version, you are deleting somebody work by pushing "old" files without latest changes. :P
Commit -  eb8876c4226265161f8a6da69ca6c62be5ec0f29

**Comment by judge2020** [comment]
I was wondering, what are the proposed benefits of updating to .NET 4.7.1 compared to the downsides?

 As of now, every Windows installation from Windows 8 includes .NET 4.5, with windows 7 getting ".NET Framework 4 Client Profile" in windows update. If 4.7.1 would be required to run HDT, based on [.net release history](https://en.wikipedia.org/wiki/.NET_Framework#Release_history), users would need to download and install the latest .NET themselves or be on the Windows 10 fall creators update v1709 which could break existing HDT installations.

**Comment by Latency** [comment]
- Opened a new PR on a different branch for this one.
- Went through all the diffs here for a second time.. everything looks right.
- GUI formatting on the xaml files looks to be reverted back too.   TESTED!!!
- Had to revamp the .csproj configuration for some of the conflicting dependencies from NuGet auto-including conflicts. 
- Perhaps, download my forked branch and test it out?

**Comment by Latency** [comment]
Hello Hunter,

 

Actually, even the fall creators build (which I run) doesn’t come with 4.7.1.   I have installed the SDK & RTs for development purposes for the IDE.

There are many changes since .NET 4.5.   You can read them on the version history for the .NET Framework and specifically for only v4.7.1 here <https://github.com/Microsoft/dotnet/tree/master/releases/net471> .

 

The work for the PR is to fix many vulnerabilities for NPE hotpoint and exception prone areas.    This is not really a necessity, but upgrading to the new compiler build configuration required that I fix the hierarchy of the namespaces for many of the files that were not in their corresponding folders.  Everything was intertwined.  The build configurations for the dependency libraries (in my PR) now support .NET Core 2.0 & .NET 4.7.1 framework for anybody who wishes to compile against them in the future.

 

In efforts to try to make the HSDT proj the same way, it really wasn’t possible.   Many of the libraries like System.Drawing and Xaml and bunches of GUI aren’t supported in Core yet.. which is why Xamarin still exists.

I spent lots of hours trying to at least use the new CLA for it, but I couldn’t get it to work.

 

What I do have now, is something in between both for the new HSDT project build configuration.   This is now reflective in todays PR release now branched & staged for release as SEPARATE from my ResourceGenerator/main one for now.

 

-Latency

**Pull request closed without merging**